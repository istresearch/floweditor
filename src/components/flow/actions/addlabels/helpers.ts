import { AddLabelsFormState } from 'components/flow/actions/addlabels/AddLabelsForm';
import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { AddLabels, Label } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): AddLabelsFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.add_input_labels) {
    const action = settings.originalAction as AddLabels;
    return {
      labels: {
        value: action.labels.map((label: Label) => {
          if (label.name_match) {
            return { name: label.name_match, expression: true };
          }
          return label;
        })
      },
      valid: true
    };
  }

  return {
    labels: { value: [] },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  formState: AddLabelsFormState
): AddLabels => {
  const result = {
    type: Types.add_input_labels,
    labels: formState.labels.value.map((label: any) => {
      if (label.expression) {
        return { name_match: label.name };
      }
      return label;
    }),
    uuid: getActionUUID(settings, Types.add_input_labels)
  };
  return result;
};
