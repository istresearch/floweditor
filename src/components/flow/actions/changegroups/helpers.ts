import { Group } from 'flowTypes';
import { Asset, AssetType } from 'store/flowContext';
import { FormState } from 'store/nodeEditor';

export const labelSpecId = 'label';

export interface ChangeGroupsFormState extends FormState {
  groups: any;
  removeAll?: boolean;
}

export const excludeDynamicGroups = (group: any): boolean => {
  return !!group.query;
};

export const mapGroupsToAssets = (groups: Group[]): Asset[] => {
  return groups.map(({ name, uuid, name_match }) => {
    if (name_match) {
      return { name: name_match, id: name_match, type: AssetType.NameMatch };
    }
    return { name, id: uuid, type: AssetType.Group };
  });
};

export const mapAssetsToGroups = (searchResults: Asset[] | null): Group[] => {
  if (!searchResults) {
    return [];
  }

  const groups = searchResults.map(result => {
    if (!result.id || result.id === result.name) {
      return {
        name_match: result.name
      };
    }
    return {
      uuid: result.id,
      name: result.name
    };
  });

  return groups;
};
