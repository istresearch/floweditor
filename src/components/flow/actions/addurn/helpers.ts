import { AddURNFormState } from 'components/flow/actions/addurn/AddURNForm';
import { getActionUUID } from 'components/flow/actions/helpers';
import { SelectOption } from 'components/form/select/SelectElement';
import { Types } from 'config/interfaces';
import { Scheme, SCHEMES } from 'config/typeConfigs';
import { AddURN } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const getSchemeOptions = (): SelectOption[] => {
  return SCHEMES.map((scheme: Scheme) => {
    return schemeToSelectOption(scheme);
  });
};

export const getSchemeObject = (scheme: string): Scheme => {
  return SCHEMES.find((item: Scheme) => item.scheme === scheme)!;
};

export const getSchemeSelectOption = (scheme: string): SelectOption =>
  schemeToSelectOption(getSchemeObject(scheme));

export const schemeToSelectOption = (scheme: Scheme): SelectOption => {
  return { value: scheme.scheme, name: scheme.path };
};

export const initializeForm = (settings: NodeEditorSettings): AddURNFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.add_contact_urn) {
    const { scheme, path } = settings.originalAction as AddURN;

    return {
      scheme: { value: getSchemeSelectOption(scheme) },
      path: { value: path },
      valid: true
    };
  }

  return {
    scheme: { value: getSchemeSelectOption('tel') },
    path: { value: '' },
    valid: false
  };
};

export const stateToAction = (settings: NodeEditorSettings, formState: AddURNFormState): AddURN => {
  return {
    type: Types.add_contact_urn,
    uuid: getActionUUID(settings, Types.add_input_labels),
    scheme: formState.scheme.value.value,
    path: formState.path.value
  };
};
