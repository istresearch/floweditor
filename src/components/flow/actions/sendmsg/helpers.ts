/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { getActionUUID } from 'components/flow/actions/helpers';
import { Attachment, SendMsgFormState } from 'components/flow/actions/sendmsg/SendMsgForm';
import { Types } from 'config/interfaces';
import { MsgTemplating, SendMsg } from 'flowTypes';
import { Asset, AssetStore, AssetType } from 'store/flowContext';
import { AssetArrayEntry, AssetEntry, NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { SelectOption } from 'components/form/select/SelectElement';
import { createUUID } from 'utils';

export const TOPIC_OPTIONS: SelectOption[] = [
  { value: 'event', label: 'Event' },
  { value: 'account', label: 'Account' },
  { value: 'purchase', label: 'Purchase' },
  { value: 'agent', label: 'Agent' }
];

export const initializeForm = (
  settings: NodeEditorSettings,
  assetStore: AssetStore
): SendMsgFormState => {
  let template: AssetEntry = { value: null };
  let templateVariables: StringEntry[] = [];

  if (settings.originalAction && settings.originalAction.type === Types.send_msg) {
    const action = settings.originalAction as SendMsg;
    const attachments: Attachment[] = [];
    (action.attachments || []).forEach((attachmentString: string) => {
      const splitPoint = attachmentString.indexOf(':');

      const type = attachmentString.substring(0, splitPoint);
      const attachment = {
        type,
        url: attachmentString.substring(splitPoint + 1),
        uploaded: type.indexOf('/') > -1
      };

      attachments.push(attachment);
    });

    if (action.templating) {
      const msgTemplate = action.templating.template;
      template = {
        value: {
          id: msgTemplate.uuid,
          name: msgTemplate.name,
          type: AssetType.Template
        }
      };
      templateVariables = action.templating.variables.map((value: string) => {
        return {
          value
        };
      });
    }

    let labels: AssetArrayEntry = { value: [] };

    if (action.labels) {
      labels = {
        value: action.labels.map(label => {
          return { id: label.uuid, name: label.name, type: AssetType.Label };
        })
      };
    }

    return {
      topic: { value: TOPIC_OPTIONS.find(option => option.value === action.topic) },
      template,
      templateVariables,
      attachments,
      message: { value: action.text },
      quickReplies: { value: action.quick_replies || [] },
      quickReplyEntry: { value: '' },
      sendAll: action.all_urns,
      labels: labels,
      valid: true
    };
  }

  return {
    topic: { value: null },
    template,
    templateVariables: [],
    attachments: [],
    message: { value: '' },
    quickReplies: { value: [] },
    quickReplyEntry: { value: '' },
    sendAll: false,
    labels: { value: [] },
    valid: false
  };
};

export const stateToAction = (settings: NodeEditorSettings, state: SendMsgFormState): SendMsg => {
  const attachments = state.attachments
    .filter((attachment: Attachment) => attachment.url.trim().length > 0)
    .map((attachment: Attachment) => `${attachment.type}:${attachment.url}`);

  let templating: MsgTemplating = null;

  if (state.template && state.template.value) {
    let templatingUUID = createUUID();
    if (settings.originalAction && settings.originalAction.type === Types.send_msg) {
      const action = settings.originalAction as SendMsg;
      if (
        action.templating &&
        action.templating.template &&
        action.templating.template.uuid === state.template.value.id
      ) {
        templatingUUID = action.templating.uuid;
      }
    }

    templating = {
      uuid: templatingUUID,
      template: {
        uuid: state.template.value.id,
        name: state.template.value.name
      },
      variables: state.templateVariables.map((variable: StringEntry) => variable.value)
    };
  }

  const result: SendMsg = {
    attachments,
    text: state.message.value,
    type: Types.send_msg,
    all_urns: state.sendAll,
    quick_replies: state.quickReplies.value,
    labels: getAsset(state.labels.value!, AssetType.Label),
    uuid: getActionUUID(settings, Types.send_msg)
  };

  if (templating) {
    result.templating = templating;
  }

  if (state.topic.value) {
    result.topic = state.topic.value.value;
  }

  return result;
};

export const getAsset = (assets: Asset[], type: AssetType): any[] => {
  if (assets) {
    return assets
      .filter((asset: Asset) => asset.type === type)
      .map((asset: Asset) => {
        return { uuid: asset.id, name: asset.name };
      });
  }

  return [];
};
