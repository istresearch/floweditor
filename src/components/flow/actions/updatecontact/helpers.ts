import { getActionUUID } from 'components/flow/actions/helpers';
import {
  CHANNEL_PROPERTY,
  LANGUAGE_PROPERTY,
  NAME_PROPERTY,
  STATUS_PROPERTY
} from 'components/flow/props';
import { Types, ContactStatus } from 'config/interfaces';
import { getTypeConfig } from 'config/typeConfigs';
import {
  Field,
  SetContactAttribute,
  SetContactChannel,
  SetContactField,
  SetContactLanguage,
  SetContactName,
  SetContactStatus
} from 'flowTypes';
import { Asset, AssetMap, AssetStore, AssetType, REMOVE_VALUE_ASSET } from 'store/flowContext';
import {
  FormState,
  NodeEditorSettings,
  StringEntry,
  SelectOptionEntry,
  FormEntry
} from 'store/nodeEditor';
import {
  CONTACT_STATUS_OPTIONS,
  CONTACT_STATUS_ACTIVE
} from 'components/flow/actions/updatecontact/UpdateContactForm';

export interface UpdateContactFormState extends FormState {
  type: Types;
  name: StringEntry;
  channel: FormEntry;
  language: FormEntry;
  status: SelectOptionEntry;
  field: FormEntry;
  fieldValue: StringEntry;
}

export const initializeForm = (
  settings: NodeEditorSettings,
  assetStore: AssetStore
): UpdateContactFormState => {
  const state: UpdateContactFormState = {
    type: Types.set_contact_name,
    valid: false,
    name: { value: '' },
    channel: { value: null },
    language: { value: null },
    status: { value: CONTACT_STATUS_ACTIVE },
    field: { value: NAME_PROPERTY },
    fieldValue: { value: '' }
  };

  if (settings.originalAction) {
    const originalType = settings.originalAction.type;
    // these have aliases, so compare the config we resolve to
    if (getTypeConfig(originalType) === getTypeConfig(Types.set_contact_field)) {
      state.type = originalType;

      switch (originalType) {
        case Types.set_contact_field:
          const fieldAction = settings.originalAction as SetContactField;
          state.field = { value: { key: fieldAction.field.key, label: fieldAction.field.name } };
          state.fieldValue = { value: fieldAction.value };
          state.valid = true;
          return state;
        case Types.set_contact_channel:
          const channelAction = settings.originalAction as SetContactChannel;
          state.field = { value: CHANNEL_PROPERTY };
          state.channel = {
            value: channelAction.channel ? channelAction.channel : REMOVE_VALUE_ASSET
          };
          state.valid = true;
          return state;
        case Types.set_contact_language:
          const languageAction = settings.originalAction as SetContactLanguage;
          state.field = { value: LANGUAGE_PROPERTY };
          state.valid = true;
          state.language = {
            value: languageAction.language
              ? {
                  iso: languageAction.language,
                  name: getLanguageForCode(languageAction.language, assetStore.languages.items)
                }
              : REMOVE_VALUE_ASSET
          };
          return state;
        case Types.set_contact_status:
          const statusAction = settings.originalAction as SetContactStatus;
          state.field = { value: STATUS_PROPERTY };
          state.valid = true;
          state.status = {
            value: CONTACT_STATUS_OPTIONS.find(o => o.value === statusAction.status)
          };
          return state;
        case Types.set_contact_name:
          const nameAction = settings.originalAction as SetContactName;
          state.valid = true;
          state.name = {
            value: nameAction.name
          };
          return state;
      }
    }
  }

  // default is updating name
  return state;
};

export const stateToAction = (
  settings: NodeEditorSettings,
  state: UpdateContactFormState
): SetContactAttribute => {
  /* istanbul ignore else */
  const field = state.field.value;
  if (state.type === Types.set_contact_field) {
    return {
      uuid: getActionUUID(settings, Types.set_contact_field),
      type: state.type,
      field: { name: field.label, key: field.key },
      value: state.fieldValue.value
    };
  } else if (state.type === Types.set_contact_channel) {
    if (state.channel.value.type === REMOVE_VALUE_ASSET.type) {
      return {
        uuid: getActionUUID(settings, Types.set_contact_channel),
        type: state.type
      } as any;
    }
    return {
      uuid: getActionUUID(settings, Types.set_contact_channel),
      type: state.type,
      channel: {
        uuid: state.channel.value.uuid,
        name: state.channel.value.name
      }
    };
  } else if (state.type === Types.set_contact_language) {
    if (state.language.value.type === REMOVE_VALUE_ASSET.type) {
      return {
        uuid: getActionUUID(settings, Types.set_contact_language),
        type: state.type
      } as any;
    }
    return {
      uuid: getActionUUID(settings, Types.set_contact_language),
      type: state.type,
      language: state.language.value.iso
    };
  } else if (state.type === Types.set_contact_status) {
    return {
      uuid: getActionUUID(settings, Types.set_contact_status),
      type: state.type,
      status: state.status.value.value as ContactStatus
    };
  } else if (state.type === Types.set_contact_name) {
    return {
      uuid: getActionUUID(settings, Types.set_contact_name),
      type: state.type,
      name: state.name.value
    };
  }
};

export const sortFieldsAndProperties = (a: any, b: any): number => {
  const aType = a.type || '';
  const bType = b.type || '';

  const aName = a.name || a.label || '';
  const bName = b.name || b.label || '';

  // Name always goes first
  /* istanbul ignore else */
  if (a.id === NAME_PROPERTY.id && b.id !== NAME_PROPERTY.id) {
    return -1;
  } else if (b.id === NAME_PROPERTY.id && a.id !== NAME_PROPERTY.id) {
    return 1;
  }

  if (aType === bType) {
    return aName.localeCompare(bName);
  }

  if (aType === AssetType.Scheme) {
    return 1;
  }

  if (bType === AssetType.Scheme) {
    return -1;
  }

  // go with alpha-sort for everthing else
  else if (aType !== bType) {
    if (aType === AssetType.ContactProperty) {
      return -1;
    }

    if (bType === AssetType.ContactProperty) {
      return 1;
    }
  }
  // non-name non-fields go last
  return aName.localeCompare(bName);
};

export const fieldToAsset = (field: Field = { key: '', name: '' }): Asset => ({
  id: field.key,
  name: field.name,
  type: AssetType.Field
});

export const assetToField = (asset: Asset): Field => ({
  key: asset.id,
  name: asset.name
});

export const assetToChannel = (asset: Asset): any => {
  if (asset.id === REMOVE_VALUE_ASSET.id) {
    return {};
  }

  return {
    uuid: asset.id,
    name: asset.name
  };
};

export const getLanguageForCode = (code: string, languages: AssetMap) => {
  let lang = code;
  if (languages && lang in languages) {
    lang = languages[lang].name;
  }
  return lang;
};

export const getName = (asset: any): string => {
  return asset.label || asset.name;
};
