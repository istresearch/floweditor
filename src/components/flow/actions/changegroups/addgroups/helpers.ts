import { ChangeGroupsFormState } from 'components/flow/actions/changegroups/helpers';
import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { ChangeGroups, Group } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): ChangeGroupsFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.add_contact_groups) {
    const action = settings.originalAction as ChangeGroups;
    return {
      groups: {
        value: action.groups.map((group: Group) => {
          if (group.name_match) {
            return { name: group.name_match, expression: true };
          }
          return group;
        })
      },
      valid: true
    };
  }

  return {
    groups: { value: null },
    valid: false
  };
};

export const stateToAction = (
  nodeSettings: NodeEditorSettings,
  state: ChangeGroupsFormState
): ChangeGroups => {
  return {
    type: Types.add_contact_groups,
    groups: state.groups.value.map((group: any) => {
      if (group.expression) {
        return { name_match: group.name };
      }
      return group;
    }),
    uuid: getActionUUID(nodeSettings, Types.add_contact_groups)
  };
};
