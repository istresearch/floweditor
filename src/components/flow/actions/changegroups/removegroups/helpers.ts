import { ChangeGroupsFormState } from 'components/flow/actions/changegroups/helpers';
import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { RemoveFromGroups } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): ChangeGroupsFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.remove_contact_groups) {
    const action = settings.originalAction as RemoveFromGroups;

    const groups = action.groups || [];
    return {
      groups: { value: groups },
      removeAll: groups.length === 0 || action.all_groups,
      valid: true
    };
  }

  return {
    groups: { value: null },
    removeAll: false,
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  state: ChangeGroupsFormState
): RemoveFromGroups => {
  return {
    type: Types.remove_contact_groups,
    groups: state.removeAll
      ? []
      : state.groups.value.map((group: any) => {
          return { uuid: group.uuid, name: group.name };
        }),
    all_groups: !!state.removeAll,
    uuid: getActionUUID(settings, Types.remove_contact_groups)
  };
};
