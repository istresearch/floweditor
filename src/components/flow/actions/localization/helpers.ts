import { KeyLocalizationFormState } from 'components/flow/actions/localization/KeyLocalizationForm';
import { MsgLocalizationFormState } from 'components/flow/actions/localization/MsgLocalizationForm';
import { Types } from 'config/interfaces';
import { getTypeConfig } from 'config/typeConfigs';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { SendMsg, MsgTemplating, SayMsg } from 'flowTypes';
import { Attachment } from '../sendmsg/attachments';

export const initializeLocalizedKeyForm = (
  settings: NodeEditorSettings
): KeyLocalizationFormState => {
  const keyValues: { [key: string]: StringEntry } = {};
  const localized = settings.localizations[0];
  const action = localized.getObject() as any;

  const keys = settings.originalAction
    ? getTypeConfig(settings.originalAction.type).localizeableKeys || []
    : [];
  keys.forEach((key: string) => {
    keyValues[key] = { value: key in localized.localizedKeys ? action[key] : '' };
  });

  return {
    keyValues,
    valid: true
  };
};

export const initializeLocalizedForm = (settings: NodeEditorSettings): MsgLocalizationFormState => {
  const state: MsgLocalizationFormState = {
    message: { value: '' },
    quickReplies: { value: [] },
    templateVariables: [],
    templating: null,
    audio: { value: null },
    valid: true,
    attachments: []
  };

  // check if our form should use a localized action
  if (
    settings.originalAction &&
    (settings.originalAction.type === Types.send_msg ||
      settings.originalAction.type === Types.say_msg) &&
    settings.localizations &&
    settings.localizations.length > 0
  ) {
    if (settings.originalAction && (settings.originalAction as any).templating) {
      state.templating = (settings.originalAction as any).templating;
      state.templateVariables = state.templating.variables.map((value: string) => {
        return {
          value: ''
        };
      });
    }

    for (const localized of settings.localizations) {
      if (localized.isLocalized()) {
        const localizedObject = localized.getObject() as any;

        if (localizedObject.text) {
          const action = localizedObject as (SendMsg & SayMsg);
          state.message.value = 'text' in localized.localizedKeys ? action.text : '';
          state.audio.value = 'audio_url' in localized.localizedKeys ? action.audio_url : null;
          state.quickReplies.value =
            'quick_replies' in localized.localizedKeys ? action.quick_replies || [] : [];

          const attachments: Attachment[] = [];

          if ('attachments' in localized.localizedKeys) {
            (action.attachments || []).forEach((attachmentString: string) => {
              const splitPoint = attachmentString.indexOf(':');

              const type = attachmentString.substring(0, splitPoint);
              const attachment = {
                type,
                url: attachmentString.substring(splitPoint + 1),
                uploaded: type.indexOf('/') > -1
              };

              attachments.push(attachment);
            });
          }

          state.attachments = attachments;
          state.valid = true;
        }

        if (localizedObject.variables) {
          const templating = localizedObject as MsgTemplating;
          state.templateVariables = templating.variables.map((value: string) => {
            return {
              value: 'variables' in localized.localizedKeys ? value : ''
            };
          });
          state.valid = true;
        }
      }
    }
  }
  return state;
};
