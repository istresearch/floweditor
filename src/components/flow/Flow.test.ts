import {
  Flow,
  FlowStoreProps,
  getDragStyle,
  ghostNodeSpecId,
  isDraggingBack,
  REPAINT_TIMEOUT
} from 'components/flow/Flow';
import { getDraggedFrom } from 'components/helpers';
import { FlowTypes } from 'config/interfaces';
import { createEmptyNode, getFlowComponents } from 'store/helpers';
import { ConnectionEvent } from 'store/thunks';
import {
  composeComponentTestUtils,
  composeDuxState,
  getSpecWrapper,
  mock,
  setMock
} from 'testUtils';
import { createUUID, merge, set } from 'utils';
import * as utils from 'utils';

jest.mock('services/Plumber');
jest.useFakeTimers();

mock(utils, 'createUUID', utils.seededUUIDs());

const {
  flowContext: { definition }
} = composeDuxState();

const { renderNodeMap: initialNodes } = getFlowComponents(definition);

const baseProps: FlowStoreProps = {
  ghostNode: null,
  debug: null,
  translating: false,
  popped: null,
  dragActive: false,
  mergeEditorState: jest.fn(),
  definition,
  nodeEditorSettings: null,
  nodes: initialNodes,
  metadata: null,
  updateConnection: jest.fn(),
  onOpenNodeEditor: jest.fn(),
  onRemoveNodes: jest.fn(),
  onUpdateCanvasPositions: jest.fn(),
  resetNodeEditingState: jest.fn(),
  onConnectionDrag: jest.fn(),
  updateSticky: jest.fn()
};

const { setup, spyOn } = composeComponentTestUtils(Flow, baseProps);

describe(Flow.name, () => {
  let ghostNodeFromWait: any;
  let ghostNodeFromAction: any;
  let mockConnectionEvent: Partial<ConnectionEvent>;

  const { nodes } = baseProps;
  const nodeMapKeys = Object.keys(nodes);

  beforeEach(() => {
    const waitNode = nodes[nodeMapKeys[nodeMapKeys.length - 6]];

    ghostNodeFromWait = createEmptyNode(
      waitNode,
      waitNode.node.exits[0].uuid,
      1,
      FlowTypes.MESSAGING
    );

    const actionNode = nodes[nodeMapKeys[0]];
    ghostNodeFromAction = createEmptyNode(
      actionNode,
      actionNode.node.exits[0].uuid,
      1,
      FlowTypes.MESSAGING
    );

    mockConnectionEvent = {
      sourceId: `${createUUID()}:${createUUID()}`,
      targetId: createUUID(),
      suspendedElementId: createUUID(),
      source: null
    };

    // Clear instance mocks
    jest.clearAllTimers();
  });

  const dragSelection = {
    startX: 270,
    startY: 91,
    currentX: 500,
    currentY: 302,
    selected: {
      '46e8d603-8e5d-4435-97dd-1333291aafca': { left: 500, top: 300 },
      'bc978e00-2f3d-41f2-87c1-26b3f14e5925': { left: 300, top: 200 }
    }
  };

  describe('helpers', () => {
    describe('isDraggingBack', () => {
      it('should return false if event indicates user is not returning to drag origin', () => {
        expect(isDraggingBack(mockConnectionEvent as ConnectionEvent)).toBeFalsy();
      });

      it('should return true if event indicates user is returning to drag origin', () => {
        const suspendedElementId = createUUID();

        expect(
          isDraggingBack({
            ...mockConnectionEvent,
            source: document.createElement('div'),
            suspendedElementId,
            targetId: suspendedElementId
          } as ConnectionEvent)
        ).toBeTruthy();
      });
    });

    describe('getDragStyle', () => {
      it('should return style object for drag selection box', () => {
        expect(getDragStyle(dragSelection)).toMatchSnapshot();
      });
    });
  });

  describe('render', () => {
    it('should render NodeEditor', () => {
      const { wrapper } = setup(true, {
        mergeEditorState: set(jest.fn()),
        nodeEditorSettings: { $set: { originalNode: null } }
      });

      expect(wrapper.find('Connect(NodeEditor)').props()).toMatchSnapshot();
    });

    it('should render Simulator', () => {
      const { wrapper } = setup(
        true,
        { mergeEditorState: set(jest.fn()) },
        {},
        {
          config: { endpoints: merge({ simulateStart: 'startMeUp' }) }
        }
      );

      expect(wrapper.find('Connect(Simulator)').props()).toMatchSnapshot();
    });
  });

  describe('instance methods', () => {
    describe('componentDidMount', () => {
      const componentDidMountSpy = spyOn('componentDidMount');
      const { instance, props } = setup(true);

      jest.runAllTimers();

      expect(componentDidMountSpy).toHaveBeenCalledTimes(1);

      expect(instance.Plumber.bind).toHaveBeenCalledTimes(7);
      expect(instance.Plumber.bind).toHaveBeenCalledWith('connection', expect.any(Function));
      expect(instance.Plumber.bind).toHaveBeenCalledWith('connection', expect.any(Function));
      expect(instance.Plumber.bind).toHaveBeenCalledWith('connectionDrag', expect.any(Function));
      expect(instance.Plumber.bind).toHaveBeenCalledWith(
        'connectionDragStop',
        expect.any(Function)
      );
      expect(instance.Plumber.bind).toHaveBeenCalledWith('beforeStartDetach', expect.any(Function));
      expect(instance.Plumber.bind).toHaveBeenCalledWith('beforeDetach', expect.any(Function));
      expect(instance.Plumber.bind).toHaveBeenCalledWith('beforeDrop', expect.any(Function));
      componentDidMountSpy.mockRestore();
    });

    describe('componenWillUnmount', () => {
      const componentWillUnmountSpy = spyOn('componentWillUnmount');
      const { wrapper, instance } = setup();

      wrapper.unmount();

      expect(componentWillUnmountSpy).toHaveBeenCalledTimes(1);
      expect(instance.Plumber.reset).toHaveBeenCalledTimes(1);

      componentWillUnmountSpy.mockRestore();
    });

    describe('onBeforeConnectorDrop', () => {
      it('should call resetNodeEditingState prop', () => {
        const { wrapper, instance, props } = setup(true, {
          resetNodeEditingState: setMock()
        });

        instance.onBeforeConnectorDrop(mockConnectionEvent);

        expect(props.resetNodeEditingState).toHaveBeenCalledTimes(1);
      });

      it('should return false if pointing to itself', () => {
        const { instance } = setup();
        const sourceId = createUUID();

        expect(
          instance.onBeforeConnectorDrop({
            ...mockConnectionEvent,
            sourceId: `${sourceId}:${createUUID()}`,
            targetId: sourceId
          })
        ).toBeFalsy();
      });

      it('should return true if not pointing to itself', () => {
        const { instance, props } = setup();
        const uuids = Object.keys(props.nodes);
        const connectionEvent = {
          sourceId: `${uuids[0]}:${props.nodes[uuids[0]].node.exits[0].uuid}`,
          targetId: uuids[1],
          suspendedElementId: uuids[0],
          source: null as any
        };

        expect(instance.onBeforeConnectorDrop(connectionEvent)).toBeTruthy();
      });
    });

    describe('onConnectorDrop', () => {
      it('should not do NodeEditor work if the user is dragging back', () => {
        const { instance, props } = setup(true, {
          onOpenNodeEditor: setMock()
        });

        jest.runAllTimers();

        instance.onConnectorDrop(mockConnectionEvent);

        expect(instance.Plumber.recalculate).not.toHaveBeenCalled();
        expect(instance.Plumber.connect).not.toHaveBeenCalled();
        expect(props.onOpenNodeEditor).not.toHaveBeenCalled();
      });

      it('should do NodeEditor work if the user is not dragging back', () => {
        // tslint:disable-next-line:no-shadowed-variable
        const { instance, props } = setup(true, {
          onOpenNodeEditor: setMock(),
          ghostNode: set(ghostNodeFromWait)
        });

        instance.onConnectorDrop({
          ...mockConnectionEvent,
          source: document.createElement('div')
        });

        jest.runAllTimers();

        expect(instance.Plumber.recalculate).toHaveBeenCalledTimes(1);
        expect(instance.Plumber.recalculate).toHaveBeenCalledWith(props.ghostNode.node.uuid);
        expect(instance.Plumber.connect).toHaveBeenCalledTimes(1);
        expect(instance.Plumber.connect).toMatchCallSnapshot();

        expect(props.onOpenNodeEditor).toHaveBeenCalledTimes(1);
        expect(props.onOpenNodeEditor).toMatchCallSnapshot();
      });
    });

    describe('beforeConnectionDrag', () => {
      it('should return reversse of translating prop', () => {
        const { wrapper, instance, props } = setup();

        expect(instance.beforeConnectionDrag(mockConnectionEvent)).toBe(!props.translating);
      });
    });
  });
});
