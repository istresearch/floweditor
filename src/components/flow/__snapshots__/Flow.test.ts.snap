// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow helpers getDragStyle should return style object for drag selection box 1`] = `
Object {
  "height": 211,
  "left": 270,
  "top": 91,
  "width": 230,
}
`;

exports[`Flow instance methods onConnectorDrop should do NodeEditor work if the user is not dragging back 1`] = `
Array [
  "46e8d603-8e5d-4435-97dd-1333291aafca:8653ac06-00ba-4890-a79f-d30f3a55193b",
  "52036e99-4be0-4505-add4-159e8d1d7572",
]
`;

exports[`Flow instance methods onConnectorDrop should do NodeEditor work if the user is not dragging back 2`] = `
Array [
  Object {
    "originalAction": Object {
      "text": "",
      "type": "send_msg",
      "uuid": "7966434b-6d38-4d8b-b01c-13d3d9aa593c",
    },
    "originalNode": Object {
      "ghost": true,
      "inboundConnections": Object {
        "8653ac06-00ba-4890-a79f-d30f3a55193b": "46e8d603-8e5d-4435-97dd-1333291aafca",
      },
      "node": Object {
        "actions": Array [
          Object {
            "text": "",
            "type": "send_msg",
            "uuid": "7966434b-6d38-4d8b-b01c-13d3d9aa593c",
          },
        ],
        "exits": Array [
          Object {
            "destination_uuid": null,
            "uuid": "431dadb3-0257-432f-ace3-52157bba69ed",
          },
        ],
        "uuid": "52036e99-4be0-4505-add4-159e8d1d7572",
      },
      "ui": Object {
        "position": Object {
          "left": 0,
          "top": 0,
        },
        "type": "execute_actions",
      },
    },
  },
]
`;

exports[`Flow render should render NodeEditor 1`] = `
Object {
  "helpArticles": Object {},
  "plumberConnectExit": [MockFunction],
}
`;

exports[`Flow render should render Simulator 1`] = `
Object {
  "mergeEditorState": [MockFunction],
  "onToggled": [Function],
  "popped": null,
}
`;
